inst/include/cpp11/r_vector.hpp

```diff
@@ -1249,6 +1274,83 @@ inline typename r_vector<T>::iterator r_vector<T>::iterator::operator+(R_xlen_t
   return it;
 }
 
+// Compared to `Rf_xlengthgets()`:
+// - This always allocates, even if it is the same size, which is important when we use
+//   it in a constructor and need to ensure that it duplicates on the way in.
+// - This copies over attributes with `Rf_copyMostAttrib()`, which is important when we
+//   use it in constructors and when we truncate right before returning from the `SEXP`
+//   operator.
+// - This is more friendly to ALTREP `x`.
+template <typename T>
+inline SEXP r_vector<T>::reserve_data(SEXP x, bool is_altrep, R_xlen_t size) {
```

Comment:
Important invariants of `reserve_data()` vs `Rf_xlengthgets()`

Useful for both:
- Usage in constructors, when we need to enforce a duplication to occur
- Usage during truncation, when we need to copy attributes too